### ISSUE TITLE
Buffer is not defined

### ISSUE THREAD

**Issue by gaoweimeng** [sequelize/sequelize#14654]
Project composition: vite+electron+sequelize
Descrietion: When I use _sequelize_ in my project and in _main.js_ When introducing models, The startup project will throw an exception of _Buffer is not defined_
Repository url: https://github.com/gaoweimeng/my-notes.git
error image:
<img width="1545" alt="image" src="https://user-images.githubusercontent.com/35355806/174245948-2456da1d-de81-4e71-8a03-6248789b017b.png">

**Comment by bluwy** [comment]
Duplicate of https://github.com/vitejs/vite/issues/8633. I think this is invalid as sequelize is advertised as a **Node.js** ORM tool, not for browser usage.

**Comment by gaoweimeng** [comment]
OK, I think there is a problem with the way I use it. I decided to use express as a middleware to operate database level data. Thank you for your answer. But _sequelize_ cannot be used in main.js?

**Comment by bluwy** [comment]
`main.js` is the entrypoint for browser code, so it can't be used.

**Comment by evanrittenhouse** [comment]
I'd think so, yes. Sequelize is a Node.js tool and is not meant to run in the browser - `buffer` is available for Node, but not the browser.

**Comment by ephys** [comment]
That's correct, Sequelize is not designed to run outside of a Node env (for now, there are requests for Deno, react-native, etc).

I'll close but feel free to continue the discussion

**Comment by catamphetamine** [comment]
> Sequelize is not designed to run outside of a Node env 

I wonder why though.
Theoretically, there shouldn't be any restrictions on why couldn't it run in a web browser.

For example, it [can work](https://github.com/sequelize/sequelize/issues/3086#issuecomment-1607318694) with an in-browser `sql.js` database.
So there's no requirement for any networking code, and still it throws `Buffer is not defined` when bundled via rollup.

I've attempted including [`rollup-plugin-polyfill-node`](https://github.com/FredKSchott/rollup-plugin-polyfill-node/) but rollup is a cryptic thing of its own, throwing strange errors depending on the order of the `plugins` in its config. When shifting plugins around, the error messages just changed but the end result was that it still didn't work in a web browser.

I'm not saying that I'd use it in a web browser. It was just a little mental excercise of mine. One example could be allowing users to play around with Sequelize in an in-browser sandbox or smth like that.

**Comment by WikiRik** [comment]
Interesting use case, it might be nice to include a sandbox for sequelize in our documentation somewhere in the future but I think that the reason mainly is that the current maintainers have  plenty ideas that they want to work on and limited interest on running sequelize outside of Node. But we're always happy to look at community PRs that can add this support.

**Comment by ephys** [comment]
Using it in the browser is something that crossed my mind many times while writing the new documentation

> I wonder why though.

Because we heavily rely on node APIs, and replacing them with isomorphic alternatives is going to require a lot of work for relatively low demand.

In my original message I did mention that this was the current state. We're not opposed to someone opening a feature requests for supporting browsers that we can then use to track interest, like was done for Deno (https://github.com/sequelize/sequelize/issues/12836)

**Comment by catamphetamine** [comment]
@WikiRik 

> But we're always happy to look at community PRs that can add this support.

@ephys

> Because we heavily rely on node APIs, and replacing them with isomorphic alternatives is going to require a lot of work for relatively low demand.

> In my original message I did mention that this was the current state. We're not opposed to someone opening a feature requests for supporting browsers

Ok, I'll create an experimental repo then.
Replacing Node.js native modules with pure-javascript "shims" doesn't seem like an impossible task. Maybe even I could do that at some point in future. Or maybe someone else wanders here and finds the fork. 

https://github.com/catamphetamine/sequelize/blob/feature/browser/packages/core/README-BROWSER.md

For now, I've just set up Rollup builder that outputs a "bundle" of Sequelize for use in a web browser. No issues were fixed, but I've played around a bit with replacing `require('crypto')` with a quick pure-javascript "shim" and I guess it ate that.
